// redaction/lib/lib/wrapReducers.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
var dispatch = void 0;
var waitList = [];

var resolveDispatch = exports.resolveDispatch = function resolveDispatch(_dispatch) {
  while (waitList.length) {
    var action = waitList.shift();
    action(_dispatch);
  }
  dispatch = _dispatch;
};

exports.default = function (fromJS) {
  return function (reducers, rootKey) {
    var dispatchedReducers = {};

    for (var nodeName in reducers) {
      if (!reducers.hasOwnProperty(nodeName)) continue;

      dispatchedReducers[nodeName] = {};

      var _loop = function _loop(methodName) {
        if (!reducers[nodeName].hasOwnProperty(methodName)) return 'continue';
        if (methodName === 'default') return 'continue';
        if (methodName === 'initialState') return 'continue';

        var type = nodeName + '.' + methodName;
        var rootType = '' + (rootKey ? rootKey + '.' : '') + type;

        var dispatchedReducer = function dispatchedReducer(payload) {
          var method = function method(dispatch) {
            return dispatch({
              type: type,
              rootType: rootType,
              payload: fromJS ? fromJS(payload) : payload
            });
          };

          if (dispatch) {
            method(dispatch);
          } else {
            waitList.push(method);
          }
        };

        dispatchedReducers[nodeName][methodName] = dispatchedReducer;
        dispatchedReducers[nodeName][methodName].type = type;
      };

      for (var methodName in reducers[nodeName]) {
        var _ret = _loop(methodName);

        if (_ret === 'continue') continue;
      }
    }

    return dispatchedReducers;
  };
};
